File: Java/src/_0059_Spiral_Matrix_II/Solution.java
package _0059_Spiral_Matrix_II;

import _0000_study._common.Output;

/**
 * Date: 2021/1/23 15:23
 * Content:
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：36.9 MB, 在所有 Java 提交中击败了5.20%的用户
 */
public class Solution {

    public int[][] generateMatrix(int n) {

        int[] dx = {-1, 0, 1, 0}, dy = {0, 1, 0, -1};
        int[][] res = new int[n][n];
        int x = 0, y = 0, d = 1;
        for (int i = 1; i <= n * n; i++) {
            res[x][y] = i;
            int a = x + dx[d], b = y + dy[d];
            if (a < 0 || a >= n || b < 0 || b >= n || res[a][b] != 0) {
                d = (d + 1) % 4;
                a = x + dx[d];
                b = y + dy[d];
            }
            x = a;
            y = b;
        }
        return res;
    }

    public static void main(String[] args) {

        int[][] res = (new Solution()).generateMatrix(3);
        Output.OutputBasicArray2D(res);
    }
}

File: Python/_0059_Spiral_Matrix_II/Solution.py
# coding=utf-8
# Date: 2021/5/24 17:04
from typing import List


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        dx = [0, 1, 0, -1]; dy = [1, 0, -1, 0]
        res = [[0 for _ in range(n)] for _ in range(n)]
        x = 0; y = 0; d = 0
        for i in range(1, n * n + 1):
            res[x][y] = i
            a = x + dx[d]; b = y + dy[d]
            if a < 0 or a >= n or b < 0 or b >= n or res[a][b] != 0:
                d = (d + 1) % 4
                a = x + dx[d]; b = y + dy[d]
            x = a; y = b
        return res


if __name__ == "__main__":
    print(Solution().generateMatrix(3))  # [[1, 2, 3], [8, 9, 4], [7, 6, 5]]
